### PR TITLE
Guard CopyUtil optional services access with workbench running

### PR DESCRIPTION
**Pull request by laeubi**
Currently activateCopy fails if no workbench is running, but all
services are optional, so it seems valid to assume that no workbench
could be handled like no services aviable.

### CODE REVIEW

**Review by mickaelistria — DISMISSED** [review]

File: bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java
Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: I have no clue, maybe @vogella knows about? At least I suspect that much more refactoring is required and given that I have no clue at all what this code provides at all....

My plan would be to first get basic things working and remove those bad patterns once not everything depends on the PlatformUI anymore...

Comment (vogella) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: IWorkbench has a method getApplication() which gives you access to MApplication or you could use EclipseContextFactory#getServiceContext(org.osgi.framework.BundleContext)} to get the root context.

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: > IWorkbench has a method getApplication() which gives you access to MApplication or you could use EclipseContextFactory#getServiceContext(org.osgi.framework.BundleContext)} to get the root context.

Sure but I can't find that these services are available in E4 at all.

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: e.g.

- org.eclipse.ui.swt.IFocusService
- org.eclipse.ui.handlers.IHandlerService

are both from `org.eclipse.ui`

> Hence why I think it's more sustainable for the project to invest efforts in identifying the bad and good patterns early to not leave room for bad patterns to multiply and stay forever.

I'm all for that, but I can't fix the past 12 years here :-) So if there is a replacement, good, but for now I'm happy I can even open the dialog from E4 at all... If we migrated all eclipse and no one missing anything we can even delete the code. I just have the feeling that there is no much move in that area and instead we more manifest the E4-Compat layer....

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: Maybe, as i have no clue what this code does it would require some help from someone else more familiar with what it actually do...

Comment (mickaelistria) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: > I have no clue what this code does

I suspect it
``` Install a copy popup menu on the specified control and activate the copy
	 * handler for the control when the control has focus. The handler will be
	 * deactivated when the control is disposed.
```

Comment (mickaelistria) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: Sorry, but we cannot accept more anti-patterns on the topic of migrating to e4. Let's do it right or do it not; everything in between is more harmful on the code itself, and also on downstream consumers that will mimic those anti-patterns.

Comment (mickaelistria) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: disabling useful (to some) features when moving to newer framework is definitely an anti-pattern as it reduces value. And it's even more obvious when there is a condition that hardcodes it.

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: > disabling useful (to some) features when moving to newer framework is definitely an anti-pattern as it reduces value.

This does not **disable** anything, as the Workbench **is** running in all cases where this code is currently able to run. it just enables a (limited) set of features for that cases where the code currently is completely **unable** to run, so this is 100% backward-compatible for the working case and adds 100% value for the case where it currently just fails as some is better than nothing.

`ICopyable` is a pure p2 interface, it is used in static context (so no e4 at all) so I have absolutely no hope in finding any "E4-like" replacement here without major effort, so if we can't decide for an intermediate solution for trivial code like this the whole improve-e4 integration story has failed for me and it seems more profitable to fork and use a custom build for that.

Funny enough in https://git.eclipse.org/r/c/equinox/rt.equinox.p2/+/190747 it was not considered an "anti-pattern" to use the exact same pattern to disable a code-path for such circumstance...

Comment (mickaelistria) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: To see it in action: Open Help > About... > Installation Defailts > Installation History; and notice that when focus is on you can right-click > Copy or Ctrl+C. This should still work with adoption of more e4. Isn't there another more e4-friendly strategy available to retrieve these services without dereferencing workbench?
And if those services cannot be accessed without dereferencing workbench, the IFocusListener can most likely be replaced by SWT FocusListeners and the IHandlerServer by EHandlerService.

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/CopyUtils.java:58]: These are good suggestions but as said beyond the scope of this change, I have created https://github.com/eclipse-equinox/p2/issues/88 now to track this with a dedicated ticket.

**Comment by mickaelistria** [comment]
>  so I'd like to merge this now so other changes are not blocked in this area...

I interpret #88 as a blocker. I really strongly dislike seeing `if (PlatformUI.isWorkbenchRunning())` in more and more places. It's really bad for the future of the Platform to spread this practice. So I stand with my -1, but this is not to be interpreted as a veto since we're both "simple" committers here.
I leave it to you how to proceed further. I think getting external people to give their opinion as you just did is a good way.

**Comment by laeubi** [comment]
@eclipse-equinox/eclipse-equinox-committers anyone?

**Pull request merged**
